### PR TITLE
[Merged by Bors] - feat: Shattering

### PR DESCRIPTION
**Pull request by YaelDillies**
Define when a finset family shatters a finset, the shatterer of a finset, define the Vapnik-Chervonenkis dimension of a finset family and prove the Sauer-Shelah lemma

---

- [x] depends on: #7522

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SetFamily/Shatter.lean [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+* Strong shattering
+-/
+
+open scoped BigOperators FinsetFamily
+
+namespace Finset
+variable {α : Type*} [DecidableEq α] {𝒜 ℬ : Finset (Finset α)} {s t : Finset α} {a : α} {n : ℕ}
+
+/-- A set family shatters a set `s` if all subsets of `s` can be obtained as the intersection of `s`
+and some element of the set family. We also say that `s` is traced by `𝒜`. -/
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shatter.lean:34]:
```suggestion
/-- A set family `𝒜` shatters a set `s` if all subsets of `s` can be obtained as the intersection of `s`
and some element of the set family, and we denote this `𝒜.Shatter s`. We also say that `s` is 
traced by `𝒜`. -/
```
(haven't checked line length)

File: Mathlib/Combinatorics/SetFamily/Shatter.lean [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+def Shatter (𝒜 : Finset (Finset α)) (s : Finset α) : Prop := ∀ ⦃t⦄, t ⊆ s → ∃ u ∈ 𝒜, s ∩ u = t
+
+instance : DecidablePred 𝒜.Shatter := fun _s ↦ decidableForallOfDecidableSubsets
+
+lemma Shatter.exists_inter_eq_singleton (hs : Shatter 𝒜 s) (ha : a ∈ s) : ∃ t ∈ 𝒜, s ∩ t = {a} :=
+  hs $ singleton_subset_iff.2 ha
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shatter.lean:40]:
```suggestion
  hs <| singleton_subset_iff.2 ha
```

File: Mathlib/Combinatorics/SetFamily/Shatter.lean [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+lemma Shatter.mono_left (h : 𝒜 ⊆ ℬ) (h𝒜 : 𝒜.Shatter s) : ℬ.Shatter s :=
+  fun _t ht ↦ let ⟨u, hu, hut⟩ := h𝒜 ht; ⟨u, h hu, hut⟩
+
+lemma Shatter.mono_right (h : t ⊆ s) (hs : 𝒜.Shatter s) : 𝒜.Shatter t := fun u hu ↦ by
+  obtain ⟨v, hv, rfl⟩ := hs (hu.trans h); exact ⟨v, hv, inf_congr_right hu $ inf_le_of_left_le h⟩
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shatter.lean:46]:
```suggestion
  obtain ⟨v, hv, rfl⟩ := hs (hu.trans h); exact ⟨v, hv, inf_congr_right hu <| inf_le_of_left_le h⟩
```

File: Mathlib/Combinatorics/SetFamily/Shatter.lean [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+lemma univ_shatter [Fintype α] : univ.Shatter s := shatter_of_forall_subset $ fun _ _ ↦ mem_univ _
+
+@[simp] lemma shatter_univ [Fintype α] : 𝒜.Shatter univ ↔ 𝒜 = univ := by
+  rw [shatter_iff, powerset_univ]; simp_rw [univ_inter, image_id']
+
+/-- The set family of sets that are shattered by `𝒜`. -/
+def shatterer (𝒜 : Finset (Finset α)) : Finset (Finset α) := (𝒜.biUnion powerset).filter 𝒜.Shatter
+
+@[simp] lemma mem_shatterer : s ∈ 𝒜.shatterer ↔ 𝒜.Shatter s := by
+  refine mem_filter.trans $ and_iff_right_of_imp $ fun h ↦ ?_
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shatter.lean:76]:
```suggestion
  refine mem_filter.trans <| and_iff_right_of_imp <| fun h ↦ ?_
```
I'm pretty sure `<|` is preferred over `$`, if I'm wrong feel free to disregard these suggestions

File: Mathlib/Combinatorics/SetFamily/Shatter.lean [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    · obtain ⟨u, hu, hsu⟩ := hs.1 ht
+      rw [mem_memberSubfamily] at hu
+      refine ⟨_, hu.1, ?_⟩
+      rw [←insert_inter_distrib, hsu, insert_erase ha]
+    · obtain ⟨u, hu, hsu⟩ := hs.2 ht
+      rw [mem_nonMemberSubfamily] at hu
+      refine ⟨_, hu.1, ?_⟩
+      rwa [insert_inter_of_not_mem hu.2, hsu, erase_eq_self]
+
+protected lemma Shatter.of_compression (hs : (𝓓 a 𝒜).Shatter s) : 𝒜.Shatter s := by
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shatter.lean:151]: What's the purpose of marking this protected?

Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/Shatter.lean]: Remnant from a time where it was (wrongly) called `Shatter.compression`

File: Mathlib/Combinatorics/SetFamily/Shatter.lean [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+lemma Shatter.card_le_vcDim (hs : 𝒜.Shatter s) : s.card ≤ 𝒜.vcDim := le_sup <| mem_shatterer.2 hs
+
+/-- Down-compressing decreases the VC-dimension. -/
+lemma vcDim_compress_le (a : α) (𝒜 : Finset (Finset α)) : (𝓓 a 𝒜).vcDim ≤ 𝒜.vcDim :=
+  sup_mono <| shatterer_compress_subset_shatterer _ _
+
+/-- The **Sauer-Shelah lemma**. -/
+lemma card_shatterer_le_sum_vcDim [Fintype α] :
+    𝒜.shatterer.card ≤ ∑ k in Iic 𝒜.vcDim, (Fintype.card α).choose k := by
+  simp_rw [←card_univ, ←card_powersetLen]
```
Comment (FLDutchmann) [inline Mathlib/Combinatorics/SetFamily/Shatter.lean:192]: #7667 renamed `powersetLen` to `powersetCard`, so you'll have to merge master manually.

File: Mathlib/Combinatorics/SetFamily/Shatter.lean [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+open scoped BigOperators FinsetFamily
+
+namespace Finset
+variable {α : Type*} [DecidableEq α] {𝒜 ℬ : Finset (Finset α)} {s t : Finset α} {a : α} {n : ℕ}
+
+/-- A set family `𝒜` shatters a set `s` if all subsets of `s` can be obtained as the intersection
+of `s` and some element of the set family, and we denote this `𝒜.Shatter s`. We also say that `s` is
+traced by `𝒜`. -/
+def Shatter (𝒜 : Finset (Finset α)) (s : Finset α) : Prop := ∀ ⦃t⦄, t ⊆ s → ∃ u ∈ 𝒜, s ∩ u = t
```
Comment (FLDutchmann) [inline Mathlib/Combinatorics/SetFamily/Shatter.lean:36]: At risk of bikeshedding, should this property be called `Shatters`? `𝒜.Shatters s` looks cleaner to my eyes.

**Comment by digama0** [comment]
Note: I have pushed an update to the lean toolchain because this PR was on a [buggy version](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Potential.20data.20loss.20from.20.60lake.20clean.60.20with.204.2E2.2E0-rc2.2F3/near/397875701) of the toolchain. **WARNING**: checking out old commits of this PR using v4.2.0-rc2 or v4.2.0-rc3 can cause `lake clean` to delete your mathlib folder! If you need to do so, make sure to delete `lakefile.olean` manually before running any `lake` commands.

**Review by b-mehta — APPROVED** [review]
Nice stuff!

bors merge

**Pull request closed without merging**